python/lsst/ip/isr/brighterFatterKernel.py

```diff
@@ -0,0 +1,413 @@
+        self.means = dict()
+        self.variances = dict()
+        self.rawXcorrs = dict()
+        self.badAmps = list()
+        self.shape = (8, 8)
+        self.gain = dict()
+        self.noise = dict()
+
+        # Things calculated from the PTC
+        self.meanXcorrs = dict()
+        self.valid = dict()
+
+        # Things that are used downstream
+        self.ampKernels = dict()
+        self.detKernels = dict()
+
+        if camera:
+            self.initFromCamera(camera, detectorId=kwargs.get('detectorId', None))
+
+        super().__init__(**kwargs)
+        self.requiredAttributes.update(['level', 'means', 'variances', 'rawXcorrs',
+                                        'badAmps', 'gain', 'noise', 'meanXcorrs', 'valid',
+                                        'ampKernels', 'detKernels'])
+
+    def updateMetadata(self, setDate=False, **kwargs):
+        """Update calibration metadata.
+
+        This calls the base class's method after ensuring the required
+        calibration keywords will be saved.
+
+        Parameters
+        ----------
+        setDate : `bool`, optional
+            Update the CALIBDATE fields in the metadata to the current
+            time. Defaults to False.
+        kwargs :
+            Other keyword parameters to set in the metadata.
+        """
+        kwargs['LEVEL'] = self.level
+        kwargs['KERNEL_DX'] = self.shape[0]
+        kwargs['KERNEL_DY'] = self.shape[1]
+
+        super().updateMetadata(setDate=setDate, **kwargs)
+
+    def initFromCamera(self, camera, detectorId=None):
+        """Initialize kernel structure from camera.
+
+        Parameters
+        ----------
+        camera : `lsst.afw.cameraGeom.Camera`
+            Camera to use to define geometry.
+        detectorId : `int`, optional
+            Index of the detector to generate.
+
+        Returns
+        -------
+        calib : `lsst.ip.isr.BrighterFatterKernel`
+            The initialized calibration.
+        """
+        self._instrument = camera.getName()
+
+        if self.level == 'AMP':
+            detector = camera[detectorId]
+            self._detectorId = detectorId
+            self._detectorName = detector.getName()
+            self._detectorSerial = detector.getSerial()
+            self.badAmps = []
+
+            for amp in detector:
+                ampName = amp.getName()
+                self.means[ampName] = []
+                self.variances[ampName] = []
+                self.rawXcorrs[ampName] = []
+                self.gain[ampName] = amp.getGain()
+                self.noise[ampName] = amp.getReadNoise()
+                self.meanXcorrs[ampName] = []
+                self.ampKernels[ampName] = []
+                self.valid[ampName] = []
+        elif self.level == 'DETECTOR':
+            for det in camera:
+                detName = det.getName()
+                self.detKernels[detName] = []
+
+        return self
+
+    def getLengths(self):
```

Comment:
I'm not sure if it's just me, but it's a bit hard to figure out what the purpose of the function is. I guess it's a helper function that gives you the total length of the kernel (from the product of its two dimensions), and also the shorter length for the case when you want to tile the matrix when doing convolutions? Also, why do you pop an observation at the end from the `nObservations` set?

Reply:
I see, `pop` is just selecting one of the `nObs` (which we assume have all the same length, right?)

Reply:
I added this in response to the comments above that I was duplicating this code in multiple places.  The issue is that we want the kernel and all components to be 2-d arrays, but those are not easily persistable in FITS.  This calculates the kernel size (17x17 for the standard 8 pixel covariance range), the covariance size (8x8 for the 8 pixel covariances), and the number of observations.  As you mention, the pop is just there to get the single number of observations; I'm using the `set` to ensure that all amplifiers have the same number of means, and raising if there's a difference.